### PR TITLE
BUG-26988 implement replace for categorical blocks

### LINKED ISSUE #26988: Bug: Series[category].replace casts incorrectly
```
dti = pd.date_range('2016-01-01', periods=3, tz='US/Pacific')
ser = pd.Series(dti)
cat = ser.astype('category')

>>> cat
0   2016-01-01 00:00:00-08:00
1   2016-01-02 00:00:00-08:00
2   2016-01-03 00:00:00-08:00
dtype: category
Categories (3, datetime64[ns, US/Pacific]): [2016-01-01 00:00:00-08:00, 2016-01-02 00:00:00-08:00,
                                             2016-01-03 00:00:00-08:00]

>>> cat.replace(cat[1], 'foo')
0    1451635200000000000
1    1451721600000000000
2    1451808000000000000
dtype: int64

```

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #26988 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

The replace functions used in the code path that caused the bug report appeared to densify the series. This PR adds a replace function to `CategoricalBlock` that takes advantage of the categorical data type when possible

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -585,6 +585,7 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
 - Fixed bug in comparison of ordered :class:`Categorical` that contained missing values with a scalar which sometimes incorrectly resulted in ``True`` (:issue:`26504`)
+- Bug in :func:`DataFrame.replace` and :func:`Series.replace` that would give unusual results on categorical data (:issue:`26988`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.25.0.rst]: "unusual" --> "incorrect"

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -2978,6 +2978,29 @@ def where(self, other, cond, align=True, errors='raise',
...
+                    categories[index] = value
+                    result.values.rename_categories(categories, inplace=True)
+            if convert:
+                return result.convert(by_item=True, numeric=False,
+                                      copy=not inplace)
+            else:
+                return result
+        else:
+            if not isna(value):
+                self.values.add_categories(value, inplace=True)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py]: shouldnt this depend on the `inplace` kwarg?

File: pandas/tests/series/test_replace.py [hunk 3]
```diff
@@ -292,6 +292,23 @@ def test_replace_categorical(self, categorical, numeric):
         expected = pd.Series(numeric)
         tm.assert_series_equal(expected, result, check_dtype=False)
 
+    def test_replace_categorical_single(self):
+        # GH 26988
```
Comment (jbrockmendel) [inline pandas/tests/series/test_replace.py]: the `copy` semantics of Categorical are weird (#26995).  Can you check that when you do `__setitem__` below it actually alters the appropriate objects (and only the appropriate ones)?

File: pandas/core/internals/blocks.py [hunk 4]
```diff
@@ -2978,6 +2978,29 @@ def where(self, other, cond, align=True, errors='raise',
                                                axis=axis, transpose=transpose)
         return result
 
+    def replace(self, to_replace, value, inplace=False, filter=None,
+                regex=False, convert=True):
+        if filter is None or not self.mgr_locs.isin(filter).any():
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py]: this block looks unrelated to the linked issue.  what case is it solving?  needs a separate test?

File: pandas/tests/series/test_replace.py (hunk 3 above)
Comment (JustinZhengBC) [reply pandas/tests/series/test_replace.py]: I can confirm that after `__setitem__`, `expected` has its second element changed but `c` doesn't, as intended

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: fixed

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (JustinZhengBC) [reply doc/source/whatsnew/v0.25.0.rst]: fixed

File: pandas/core/internals/blocks.py (hunk 4 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: The first case is used when the replace function applies to the entire block, so the replace can be carried out just by manipulating categories. If there's a filter, this might not be possible so we fix the categories then use the original implementation. Most calls to `Series.replace` should be covered with this case (including the new test in `series/test_replace.py`)

I added a test in `frame/test_replace.py` to cover the second case, where a filter is used. It improves upon the previous behaviour, which would have casted the frame to `object`

**Comment by JustinZhengBC** [comment]
Failure unrelated #26842

File: pandas/core/internals/blocks.py [hunk 5]
```diff
@@ -2978,6 +2978,30 @@ def where(self, other, cond, align=True, errors='raise',
                                                axis=axis, transpose=transpose)
         return result
 
+    def replace(self, to_replace, value, inplace=False, filter=None,
+                regex=False, convert=True):
+        result = self if inplace else self.copy()
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: so rather than do this here, can you add a ``.replace()`` method on ``Categorical`` and just call it here (this by-definition is a 1-D block holder).

File: pandas/tests/frame/test_replace.py [hunk 6]
```diff
@@ -1149,3 +1149,20 @@ def test_replace_method(self, to_replace, method, expected):
...
+    ])
+    def test_categorical_replace_with_dict(self, replace_dict, final_data):
+        # GH 26988
+        df = DataFrame([[1, 1], [2, 2]], columns=['a', 'b'], dtype='category')
+        expected = DataFrame(final_data, columns=['a', 'b'], dtype='category')
+        expected['a'].cat.set_categories([1, 2, 3], inplace=True)
+        expected['b'].cat.set_categories([1, 2, 3], inplace=True)
+        result = df.replace(replace_dict, 3)
+        assert_frame_equal(result, expected)
+        with pytest.raises(AssertionError):
```
Comment (jreback) [inline pandas/tests/frame/test_replace.py]: can you add a comment on what the error condition is, and why would it be an AssertionError?

File: pandas/tests/frame/test_replace.py [hunk 7]
```diff
@@ -1149,3 +1149,20 @@ def test_replace_method(self, to_replace, method, expected):
...
+
+    @pytest.mark.parametrize("replace_dict, final_data", [
+        ({'a': 1, 'b': 1}, [[3, 3], [2, 2]]),
+        ({'a': 1, 'b': 2}, [[3, 1], [2, 3]])
+    ])
+    def test_categorical_replace_with_dict(self, replace_dict, final_data):
+        # GH 26988
+        df = DataFrame([[1, 1], [2, 2]], columns=['a', 'b'], dtype='category')
+        expected = DataFrame(final_data, columns=['a', 'b'], dtype='category')
+        expected['a'].cat.set_categories([1, 2, 3], inplace=True)
```
Comment (jreback) [inline pandas/tests/frame/test_replace.py]: try not to use inplace in tests, its non-idiomatic.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: hold off on this pending discussion about what belongs on EAs

**Review by TomAugspurger — APPROVED** [review]
Fixed the merge conflict. LGTM.

**Comment by jbrockmendel** [comment]
Are there tests for the non-None `filter` branch?

**Comment by JustinZhengBC** [comment]
@jbrockmendel yes, the dataframe tests take that branch while the series tests do not

File: pandas/core/internals/blocks.py (hunk 5 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: I am -1 on this PR as is; this should be a method on a Categorical.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> I am -1 on this PR as is; this should be a method on a Categorical.

If it goes on Categorical, it should be generalized to EA.

**Comment by jreback** [comment]
> > I am -1 on this PR as is; this should be a method on a Categorical.
> 
> If it goes on Categorical, it should be generalized to EA.

probably. But I think its *much* better then jamming things in the blocks (now we *already* have replace on the blocks, this PR might be ok actually), but I don't fully remember.

**Comment by TomAugspurger** [comment]
I don't think we (or downstream EA authors) can reliably implement ExtensionArray.replace since it's such a complicated method.

We're also relying on Block.replace to do the heavy lifting. An ExtensionArray.replace wouldn't have that luxury.

**Comment by jreback** [comment]
can you merge master

File: doc/source/whatsnew/v0.25.0.rst [hunk 8]
```diff
@@ -952,6 +952,7 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
 - Fixed bug in comparison of ordered :class:`Categorical` that contained missing values with a scalar which sometimes incorrectly resulted in ``True`` (:issue:`26504`)
+- Bug in :func:`DataFrame.replace` and :func:`Series.replace` that would give incorrect results on categorical data (:issue:`26988`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: move the note to 1.0

File: pandas/core/internals/blocks.py [hunk 9]
```diff
@@ -3005,6 +3005,30 @@ def where(
             )
         return result
 
+    def replace(
+        self, to_replace, value, inplace=False, filter=None, regex=False, convert=True
+    ):
+        result = self if inplace else self.copy()
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: need

        inplace = validate_bool_kwarg(inplace, "inplace")

File: pandas/core/internals/blocks.py [hunk 10]
```diff
@@ -3005,6 +3005,30 @@ def where(
...
+        if filter is None:
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: what case hits this?

File: pandas/core/internals/blocks.py [hunk 11]
```diff
@@ -3005,6 +3005,30 @@ def where(
...
+            categories = result.values.categories.tolist()
+            if to_replace in categories:
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can we not define a Categorical.replace to handle this code here (from 3013 thru 3020)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_replace.py (hunk 6 above)
Comment (JustinZhengBC) [reply pandas/tests/frame/test_replace.py]: fixed

File: pandas/core/internals/blocks.py (hunk 10 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: filter is none when replace is called on a Series as opposed to a DataFrame

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: fixed

File: pandas/core/internals/blocks.py (hunk 11 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: done, with tests added in `pandas/tests/arrays/categorical/test_algos.py`

File: pandas/core/arrays/categorical.py [hunk 12]
```diff
@@ -2494,6 +2494,20 @@ def isin(self, values):
         code_values = code_values[null_mask | (code_values >= 0)]
         return algorithms.isin(self.codes, code_values)
 
+    def replace(self, to_replace, value, inplace=False):
+        inplace = validate_bool_kwarg(inplace, "inplace")
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: cool, can you type & add a doc-string?

File: pandas/core/arrays/categorical.py [hunk 13]
```diff
@@ -2494,6 +2494,20 @@ def isin(self, values):
...
+        cat = self if inplace else self.copy()
+        categories = cat.categories.tolist()
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: we shouldn't need to use .tolist() here (very inefficient) as its already an Index

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @jbrockmendel if any more comments

**Comment by jbrockmendel** [comment]
will take a look this afternoon

File: doc/source/whatsnew/v1.0.0.rst [hunk 14]
```diff
@@ -88,6 +88,7 @@ Categorical
 ^^^^^^^^^^^
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
+- Bug in :func:`DataFrame.replace` and :func:`Series.replace` that would give incorrect results on categorical data (:issue:`26988`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.0.0.rst]: "func" -> "meth"

File: pandas/core/arrays/categorical.py [hunk 15]
```diff
@@ -2494,6 +2494,20 @@ def isin(self, values):
...
+        if to_replace in categories:
+            if isna(value):
+                cat.remove_categories(to_replace, inplace=True)
+            else:
+                index = categories.index(to_replace)
+                categories[index] = value
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py]: what if `value` is already in categories?

**Comment by jbrockmendel** [comment]
Only major question is whether we need to add `replace` to the EA interface.

File: pandas/core/arrays/categorical.py [hunk 16]
```diff
@@ -2494,6 +2494,20 @@ def isin(self, values):
         code_values = code_values[null_mask | (code_values >= 0)]
         return algorithms.isin(self.codes, code_values)
 
+    def replace(self, to_replace, value, inplace=False):
+        inplace = validate_bool_kwarg(inplace, "inplace")
+        cat = self if inplace else self.copy()
+        categories = cat.categories.tolist()
+        if to_replace in categories:
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py]: what happens if `to_replace` is not hashable?  I think at the Categorical level we are OK raising, but at the Block level we would cast to object right?

**Comment by jreback** [comment]
this looked pretty good, can you merge master and address any remaining comments.

**Comment by jreback** [comment]
can you merge master

File: pandas/core/arrays/categorical.py (hunk 16 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: Would casting to object help?
```
>>> import pandas as pd
>>> s=pd.Series([1,2]).astype(object)
>>> s.replace(1,[1,2,3])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/series.py", line 4303, in replace
    method=method,
  File "pandas/core/generic.py", line 6819, in replace
    raise TypeError(msg)  # pragma: no cover
TypeError: Invalid "to_replace" type: 'int'
```

File: pandas/core/arrays/categorical.py (hunk 15 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: fixed with new case

File: doc/source/whatsnew/v1.0.0.rst (hunk 14 above)
Comment (JustinZhengBC) [reply doc/source/whatsnew/v1.0.0.rst]: fixed

File: pandas/core/arrays/categorical.py (hunk 13 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: fixed, toList is not used unless it is needed

File: pandas/core/arrays/categorical.py (hunk 12 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: fixed (only added types for the boolean arguments in both replace functions, do the `to_replace` and `value` parameters need types?)

File: pandas/core/arrays/categorical.py (hunk 16 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: @jreback I've merged master and fixed all comments except this one for the reason above

**Comment by jbrockmendel** [comment]
@JustinZhengBC can you merge master.  itd be nice to get this in

**Comment by JustinZhengBC** [comment]
@jbrockmendel done

File: pandas/core/arrays/categorical.py [hunk 17]
```diff
@@ -2471,6 +2471,51 @@ def isin(self, values):
         code_values = code_values[null_mask | (code_values >= 0)]
         return algorithms.isin(self.codes, code_values)
 
+    def replace(self, to_replace, value, inplace: bool = False):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you type these more specifically (can be a followon PR)

File: pandas/tests/frame/test_replace.py (hunk 6 above)
Comment (jreback) [reply pandas/tests/frame/test_replace.py]: wait, this is *really* odd, why is an AssertionError here?

**Review by jreback — COMMENTED** [review]
question about an error condition

**Comment by JustinZhengBC** [comment]
@jreback how should parameters like `to_replace` and `value` be typed when they can take on any type and might not even be the same type?

**Comment by jreback** [comment]
> @jreback how should parameters like `to_replace` and `value` be typed when they can take on any type and might not even be the same type?

right, but they are something like Union[Scalar, AnyArrayLike] I think?  (again this is likely tricky so should do as a followup)

**Comment by TomAugspurger** [comment]
Fixed the merge conflict.

I think we decided to leave typing as a followup, so this should be good to go?

File: pandas/core/internals/blocks.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: I still find this non-obvious.  can you add a comment to this effect

File: pandas/core/internals/blocks.py [hunk 18]
```diff
@@ -2924,6 +2924,30 @@ def where(
...
+            if convert:
+                return result.convert(numeric=False, copy=not inplace)
+            else:
+                return result
+        else:
+            if not isna(value):
+                result.values.add_categories(value, inplace=True)
+            return super(CategoricalBlock, result).replace(
+                to_replace, value, inplace, filter, regex, convert
+            )
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py]: can any of the logic in this method generalize to ExtensionBlock?  (presumably we'd need to add `replace` to the EA interface?)

Comment (jreback) [reply pandas/core/internals/blocks.py]: this is a good point, but we can certainly do as a followup; can you open an issue for discussion.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@JustinZhengBC if you can address @jbrockmendel comments to add some comments this lgtm and we can merge. (alternatvely @jbrockmendel could do as a followup)

File: pandas/core/internals/blocks.py (hunk 10 above)
Comment (JustinZhengBC) [reply pandas/core/internals/blocks.py]: i added a comment

**Comment by JustinZhengBC** [comment]
@jreback @jbrockmendel I added the comment

**Comment by alimcmaster1** [comment]
Seems like this is good to merge.
CC. @jreback @jbrockmendel

**Comment by jreback** [comment]
thanks for the patch @JustinZhengBC very nice!
